### ISSUE TITLE
Strip leading spaces in QHP file

### ISSUE THREAD

**Issue by ferdymercury** [doxygen/doxygen#9058]
**Describe the bug**
qhelpgenerator crashes with some doxygen-generated qhp files on big projects with a lot of nested classes. This is because there are lines in the file that are too long (12.000 indentation spaces before the content in the line starts, and that is repeated on every line).

**Expected behavior**
It would be nice to have a Doxyfile flag to strip leading spaces after generating the index.qhp file, before running qhelpgenerator. Even if it makes the file less correct from the indentation point of view.

**Screenshots**
```
qhelpgenerator index.qhp
       Error in line 134919: Premature end of document.
# Let us strip spaces, see https://stackoverflow.com/questions/59715196/how-can-i-efficiently-remove-unnecessary-whitespace-from-xml-files-with-a-linux
sed -i -e '/^ */s///' -e '/ *$/s///' index.qhp
qhelpgenerator index.qhp
       ...
       Documentation successfully generated.
```

**To Reproduce**
Download [index.qhp.gz](https://github.com/doxygen/doxygen/files/7912928/index.qhp.gz) and run qhelpgenerator on it.

**Version**
Doxygen 1.9.4 (99482978eb791a6159c4e409c42ec95acbbd2c1d*)
Qt Help Generator version 1.0 (Qt 4.8.7)

**Additional context**
This documentation is generated within the ROOT project of CERN.

**Comment by albert-github** [comment]
The fix itself would not be to difficult and the size gain can be quite big (the original file has a size of 4,989,272,510 bytes and when running the `sed` the resulting file has a size of 54,170,110 bytes).
When running the `qhelpgenerator` on the original file we get an error whilst when running on the reduced file there is no error though when trying to view the file in the assistant (version 6.1.2) it doesn't show the result (might be that there are still some files missing as I only have the qhp file).
The readability of the resulting file after the `sed` is not the best either.

I'm a bit reluctant to remove the spaces. 

@doxygen what is your opinion?

**Comment by albert-github** [comment]
Maybe a way would be to switch `GENERATE_QHP` from a boolean to a variable with 3 values:
- `NO`, no generation it is like it is now
- `YES`, generation like it is done now.
- `CONDENSED`, like `YES` but removing the leading spaces.

@doxygen what do you think?

**Comment by doxygen** [comment]
@albert-github I would make a debug option (`-d qhp`) to enable the non-condensed output. Rationale: the file is meant for processing by tools, not for humans to read. I also don't think the current indenting is correct, there seems to be some missing cases where the indent should decrease. The following example shows members of `TApplication` yet the indentation increases 4 times.

```.xml
                      <section title="fIsChecked" ref="classROOT_1_1Experimental_1_1Internal_1_1RResultBase.html#a27e82995641a3af8aa5ad7120f2bbc41" />
                      <section title="TApplication" ref="classTApplication.html">
                        <section title="EExitOnException" ref="classTApplication.html#a2f699906b45ca8bd6404e2d0cc54658b">
                          <section title="kDontExit" ref="classTApplication.html#a2f699906b45ca8bd6404e2d0cc54658ba79976c51b779212ca7227340001f7dfe" />
                          <section title="kExit" ref="classTApplication.html#a2f699906b45ca8bd6404e2d0cc54658baa261b49ff157536d21bc5164db0ca7fc" />
                          <section title="kAbort" ref="classTApplication.html#a2f699906b45ca8bd6404e2d0cc54658ba7e71cb99bd0e852d0539418b80bedf4e" />
                          <section title="EFileBits" ref="classTApplication.html#a9cdab6aad3ca84066f168c04aae56bfc">
                            <section title="kExpression" ref="classTApplication.html#a9cdab6aad3ca84066f168c04aae56bfca101cd6dd0fcf41dc84dddef2c21cce78" />
                            <section title="EStatusBits" ref="classTApplication.html#a4153a52a08881b29a3b500e63b57cff4">
                              <section title="kProcessRemotely" ref="classTApplication.html#a4153a52a08881b29a3b500e63b57cff4a44c307e438d431b39e078b8171e2245d" />
                              <section title="kDefaultApplication" ref="classTApplication.html#a4153a52a08881b29a3b500e63b57cff4ac9b451860b304bb4b204cc6391e2ff03" />
                              <section title="TApplication" ref="classTApplication.html#a2f0237bc6d078165175a97d0072fe178" />
```

**Comment by albert-github** [comment]
- OK. I think it is a good idea to use the `-d qhp` for this.
- I'll also have a look whether or not the  enum is the cause of  "strange" indentations

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #9068 (implementing the compressed form).

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #9069 (regarding the wrong indentation in case of enums)

**Comment by albert-github** [comment]
Code, of both pull requests, has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by ferdymercury** [comment]
Thanks for the fix. We have now compiled doxygen with git master.

Doxygen version used: 1.9.4 (bba8351f85981ddd950053306737b2b5177fb470)

However, we get now a different type of error:

```
Error in line 1: Opening and ending tag mismatch.
error: failed to run qhelpgenerator on index.qhp
```

Find attached the new qhp file.
[index.qhp.gz](https://github.com/doxygen/doxygen/files/7941553/index.qhp.gz)

**Comment by albert-github** [comment]
Might be that this is a problem with some structs. I have to dive into it properly.
A quick workaround would be in index.cpp move the line (around 334):
```
 Doxygen::indexList->closeContentsItem();
```
to before the `}` just before it,

I'm not happy with the differences between e.g. the handling of "section" in the qhp and the chm versions, I to have to see whether the chm doesn't have problems like qhp had. Maybe some refactoring of the qhp writing file would be appropriate.

**Comment by ferdymercury** [comment]
Hmm yes, this would maybe explain why I was seeing some weird effects when loading the help in the qt-assistant. For example, with the attached qch file, when loaded, I don't see everything under the same tree:
![image](https://user-images.githubusercontent.com/10653970/151464745-73004fa5-54d4-46d9-a2fb-b84179e41941.png)

Tutorials and Functional Parts should be under "ROOT".

**Comment by albert-github** [comment]
Indeed this is also one of the problems I do see (didn't mention it yet) and the reason for my remark "I'm not happy with the differences between e.g. the handling of "section" in the qhp and the chm versions, I to have to see whether the chm doesn't have problems like qhp had. Maybe some refactoring of the qhp writing file would be appropriate.".

**Comment by ferdymercury** [comment]
Thanks for the clarification. I think that the problem might be with the classes itself. For example, here the classes appear as recursively nested, even if they should be all at the same level. (TAttBox, TAttBox2D, TAttFill). Independently on whether there are structs or enums within.

![image](https://user-images.githubusercontent.com/10653970/151670347-e0133a1c-d230-4f83-81a2-56b0fc479ca5.png)

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
